interfaces/attrs.go

```diff
@@ -0,0 +1,147 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+type Attributes interface {
+	SetStaticAttr(key string, value interface{})
+	StaticAttr(key string) (interface{}, error)
+	StaticAttrs() map[string]interface{}
+	Attr(key string) (interface{}, error)
+	Attrs() (map[string]interface{}, error)
+	SetAttr(key string, value interface{}) error
+}
+
+type plugSlotData struct {
+	data         *snap.PlugSlotData
```

Comment:
sharing data with the original like this is probably not useful, misleading. As I wrote in one comment the point is more that the repo should be moved to be *Data based.

Reply:
we should treat the *Info stuff as immutable, and deal with repos/interfcaces needing to do mutations by using copies, the new *Data stuff